addons/udes_stock/models/stock_picking.py

```diff
@@ -739,3 +740,268 @@ def raise_stock_investigation(self, reason, quants, location):
         # Try to re-assign the picking after creating the stock investigation
         # picking as we have reserved the problematic stock
         self.action_assign()
+
+    def construct_package_hierarchy_links(self):
+        # TODO: add link/unlink to parent if needed like old _set_u_result_parent_package_id()
+        # PackageHierarchyLink = self.env["package.hierarchy.link"]
+        # new_link = PackageHierarchyLink.create(
+        #     {
+        #         "child_id": result_package.id,
+        #         "parent_id": result_parent_package.parent_id.id,
+        #         "move_line_ids": [(6, 0, mls.ids)],
+        #     }
+        # )
+        super().construct_package_hierarchy_links()
+
+    def _check_entire_pack(self):
+        """
+        Override to avoid values at result_package_id when
+        user scans products/avoid _check_entire_pack of products
+        """
+        pickings = self.filtered(
+            lambda p: p.picking_type_id.u_user_scans != "product"
+            and p.picking_type_id.u_target_storage_format != "product"
+        )
+        super(StockPicking, pickings)._check_entire_pack()
+
+    def log_name(self):
+        """This can probably go to models later"""
+        return f"Picking({self.id, self.name}"
+
+    def assert_valid_state(self):
+        """Checks if the transfer is in a valid state, i.e., not done or cancel
+        otherwise it raises and error
+        """
+        self.ensure_one()
+        if self.state in ("done", "cancel"):
+            raise ValidationError(_("Wrong state %s for %s") % (self.state, self.log_name()))
+
+    def _action_done(self):
+        """Extend to add logging"""
+        with self.statistics() as stats:
+            res = super()._action_done()
+        _logger.info(
+            "%s udes_stock::_action_done in %.2fs, %d queries", self.log_name(), stats.elapsed, stats.count
+        )
+        return res
+
+    def add_unexpected_parts(self, product_quantities):
+        """ TODO not needed for now, add to readme when implemented
+        """
+        pass
+
+    def _prepare_result_packages(
+        self,
+        product_ids,
+        package,
+        result_package_name,
+        result_parent_package_name,
+        target_storage_format=None,
+    ):
+        """
+        Compute result_package and result_parent_package based on the u_target_storage_format of the
+        picking type (if `target_storage_format` is not supplied) + the input parameters.
+        """
+        Package = self.env["stock.quant.package"]
+
+        result_parent_package = None
```

Comment:
Should some of this really be on the package model?  Picking should tell Package what format it wants, but Package decides how to create an instance of itself?

Reply:
We have 2 packages, you mean one of them or just Package in generic?

Reply:
In generic - ideally logic about creating packages would be the responsibility of Package rather than Picking.  This method is about 80 lines long, but `self` is only referenced three times - that suggests to me some of the logic might belong elsewhere :)